### PR TITLE
Reuse 'forloop' hash to save memory allocations

### PR DESCRIPTION
**Pull request by fw42**
Tries to address https://github.com/Shopify/liquid/issues/678.

Not sure if this makes sense or if there is some side-effects that I'm missing (but if I am, then at least this PR should reveal some missing test coverage, since everything is green).

Before:
```
                          user     system      total        real
ruby_empty_loop        0.030000   0.000000   0.030000 (  0.026284)
for_loop_no_body       0.840000   0.010000   0.850000 (  0.853402)
for_loop_with_body     1.800000   0.000000   1.800000 (  1.807363)
```

After:
```
                           user     system      total        real
ruby_empty_loop        0.030000   0.000000   0.030000 (  0.027673)
for_loop_no_body       0.430000   0.000000   0.430000 (  0.430292)
for_loop_with_body     1.240000   0.010000   1.250000 (  1.259245)
```

(See https://github.com/Shopify/liquid/issues/678 for the benchmark script I'm using.)

The idea here is that I don't see any reason why consecutive loop iterations each create a new Hash Ruby object for the `forloop` variable. Instead, re-use the one from the previous iteration and just update the keys (if necessary).

Also refactor the `For#render` method a little bit to (hopefully) make it a bit more readable, it was getting pretty long and hairy.

@dylanahsmith @pushrax

### CODE REVIEW

**Comment by pushrax** [comment]
Good call. I don't see any reason why this would be an issue, as it makes no sense for people to store things on this hash (let alone expect the stored value to go away in the next iteration). LGTM.

File: lib/liquid/tags/for.rb [hunk 1]
```diff
@@ -170,6 +114,74 @@ def strict_parse(markup)
...
+
+          segment.each_with_index do |item, index|
+            context[@variable_name] = item
+
+            loop_vars['index'.freeze] = index + 1
+            loop_vars['index0'.freeze] = index
+            loop_vars['rindex'.freeze] = length - index
+            loop_vars['rindex0'.freeze] = length - index - 1
+            loop_vars['first'.freeze] = (index == 0)
+            loop_vars['last'.freeze] = (index == length - 1)
```
Comment (dylanahsmith) [inline lib/liquid/tags/for.rb]: Should we instead use a liquid drop? Then we can just increment a single index between iterations of the loop, and calculate these values lazily on methods on the drop.

**Comment by dylanahsmith** [comment]
There is the possibility of a side effect from the `forloop` hash being stored in another variable, but that seems unlikely.

File: lib/liquid/tags/for.rb (hunk 1 above)
Comment (fw42) [inline lib/liquid/tags/for.rb]: hm interesting idea, I will try that

Comment (pushrax) [inline lib/liquid/tags/for.rb]: This is a good idea too.

**Comment by fw42** [comment]
Updated. Even a bit faster now:
```
                           user     system      total        real
ruby_empty_loop        0.030000   0.000000   0.030000 (  0.028785)
for_loop_no_body       0.340000   0.000000   0.340000 (  0.349778)
for_loop_with_body     1.170000   0.000000   1.170000 (  1.169294)
```

**Comment by pushrax** [comment]
Awesome stuff. My only concern with the drop approach is that some custom tags might try to access these indexes with the `[]` operator; is this something that might be the case?

**Comment by fw42** [comment]
> My only concern with the drop approach is that some custom tags might try to access these indexes with the [] operator

That should work, `Liquid::Drop` implements `#[]`.

**Comment by pushrax** [comment]
Also, this is probably a good time to remove the duplication in `TableRow`.

**Comment by pushrax** [comment]
> That should work, `Liquid::Drop` implements `#[]`.

Right :facepalm:

File: lib/liquid/forloop_drop.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    def initialize(name, length, parentloop)
+      @name = name
+      @length = length
+      @parentloop = parentloop
+      @index = 0
+    end
+
+    attr_reader :name, :length, :parentloop
+
+    def increment
```
Comment (dylanahsmith) [inline lib/liquid/forloop_drop.rb]: Do we want to make this protected so it isn't invokable from liquid?

Comment (pushrax) [inline lib/liquid/forloop_drop.rb]: :+1:

**Comment by dylanahsmith** [comment]
:+1:

File: lib/liquid/forloop_drop.rb (hunk 2 above)
Comment (fw42) [inline lib/liquid/forloop_drop.rb]: hm if I make it protected, I can't call it from the `For` class either though :-/

Comment (fw42) [inline lib/liquid/forloop_drop.rb]: I'm just gonna ghetto `send` it

**Comment by JulienItard** [comment]
any idea when it will be released with V4 ?

**Pull request merged**